### ISSUE TITLE
Linter does not flag invalid string concatenations

### ISSUE THREAD

**Issue by TheCleric** [sqlfluff/sqlfluff#6616]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When doing a string concatenation via a `+` sign in dialects that do not support it (such as spark and databricks), sqlfluff fails to flag this as invalid SQL.

### Expected Behaviour

sqlfluff will flag invalid string concatenations, and potentially guide the user to use the correct operator instead (such as `||` in databricks) or fix if desired.

I know this would be difficult to impossible for complex uses cases, but I feel like this should at least occur for simple use cases (i.e., a string constant `+` something else).

### Observed Behaviour

sqlfluff produces no warnings or errors on invalid string concatenations using a `+` on dialects that do not support it.

### How to reproduce

The following minimal reproducible case produces no errors in sqlfluff for the databricks dialect:

```sql
SELECT 'a' + 'b'
```

Whereas actually executing it in a Databricks console returns the following error:

```
[DATATYPE_MISMATCH.BINARY_OP_WRONG_TYPE] Cannot resolve "(a + b)" due to data type mismatch: the binary operator requires the input type ("NUMERIC" or "INTERVAL DAY TO SECOND" or "INTERVAL YEAR TO MONTH" or "INTERVAL"), not "STRING". SQLSTATE: 42K09; line 1, pos 7
```

### Dialect

databricks

### Version

sqlfluff, version 3.3.0
Python 3.10.13

### Configuration

```toml
[sqlfluff]
verbose = 1
dialect = databricks
max_line_length = 120
exclude_rules = CP02, ST06
warnings = RF03
large_file_skip_byte_limit = 0

[sqlfluff:indentation]
tab_space_size = 2
indented_joins = True
indented_ctes = True
indented_using_on = True
indented_on_contents = True
# indented_then = True
# indented_then_contents = True
allow_implicit_indents = True
template_blocks_indent = True

[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = upper

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = upper

[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = upper
```

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by TheCleric** [comment]
Based on this [comparison chart](https://en.wikibooks.org/wiki/SQL_Dialects_Reference/Functions_and_expressions/String_functions) I wonder if `+` should be flagged by fault on all sql and enabled for only dialects that support is (such as tsql).

**Comment by TheCleric** [comment]
It looks like this has been setup correctly in the dialects in the sense that `ConcatSegment` in ansi and its derivatives are set to be `||` and in TSQL it is `+`.

However as best as I can tell, that `ConcatSegment` is rolled up into `StringBinaryOperatorGrammar` which is then rolled up into `BinaryOperatorGrammar` where it's just used kind of generically in `Expression_A_Grammar`. `StringBinaryOperatorGrammar` is used in `Expression_B_Grammar` as a `OneOf` to join multiple together, but it's not really being used especially for strings.

I validated this by adding tests for things like `SELECT 'a' * 'b' AS ab` and these pass indicating we're not really separating out arithmetic/binary operators from string operators in the final parsing. So to fix this I think `Expression_A_Grammar` and `Expression_B_Grammar` would need to get quite a bit more complex to chase down strings vs. non-strings.

Any thoughts?

**Comment by peterbud** [comment]
@TheCleric - thanks for opening this issue.

I think each dialect should handle it's concatenation operators properly.

As I understand databricks is not yet handling this - so it should be fixed.

But I cannot follow the second comment on binary operators on strings: it seems to me as a separate issue and not related ot the databricks concatenation - can you confirm?

**Comment by TheCleric** [comment]
Trying to recreate my thinking here since this was a bit ago. 🤣

I believe what I was trying to communicate is that at a fundamental level this is handled at the `Expression_A_Grammar` level which says (in dialect_ansi, the root dialect) that you can have (among others) a `BinaryOperatorGrammar` and a `Tail_Recurse_Expression_A_Grammar`. Where a `BinaryOperatorGrammar` is one of `ArithmeticBinaryOperatorGrammar`, `StringBinaryOperatorGrammar`, `BooleanBinaryOperatorGrammar`, or `ComparisonOperatorGrammar`. Essentially this all means for a value you can have a left hand side, an operator, and a right hand side. So `a + b` or `a || b` or `a >= b`.

My problem in trying to untangle this is that the current parser isn't too particular about what's in the left and right hand side when it comes to strings. So even though the string concatenation symbols are setup correctly per dialect, the current parser sees something like `'a' + 'b'` and parses it to say "I see a value, a binary operator (specifically an arithmetic operator), and a value. So this is valid."

To get this working the "right" way, we'd need to fundamentally change the `Expression_A_Grammar` to say for non-string types this is true, but for string types you can't use arithmetic operators (or booleans as well probably). But this sort of fundamental change would have to be well tested as one wrong move could break nearly everything in the library as nearly every SQL statement in every dialect uses the ansi `Expression_A_Grammar` as far as I can tell.

As well this would still only be able to catch known strings. So even then an expression like `a + b` could be invalid (if a and b are strings) but sqlfluff may not know that a and b are strings.

**Comment by peterbud** [comment]
Thanks @TheCleric , this explanation helps. I think this gets to the core limitation here: validating expressions like `a + b` would require **semantic** type analysis, not just parsing. In many cases SQLFluff simply does not know whether `a` and `b` are strings, numerics, intervals, etc., unless it becomes schema-aware and does much deeper analysis than it does today.

Because of that, I don’t think the parser can realistically solve this in the general case. What it can do, and what seems reasonable for SQLFluff, is catch the obvious cases where we can infer intent from syntax alone, for example string literals such as 'a' + 'b', or possibly other cases adjacent to a literal, in dialects where `+` is not a valid string concatenation operator.

So my view is:

* full detection of invalid string concatenation is out of scope for a linter/parser-only approach
* a targeted rule for clear literal-based cases is in scope and seems like the pragmatic fix
* changing Expression_A_Grammar to enforce typing rules globally would be high-risk and still would not solve cases like `a + b`

**Comment by TheCleric** [comment]
I agree. I don't think we could flag this problem holistically, but even flagging the obvious (see my minimal reproducible example above) would be better than what occurs now, as long as it can be implemented sanely.

**Comment by peterbud** [comment]
Hi @TheCleric , after reading through the thread, it seems this rule idea is essentially attempting to perform a syntax check. SQLFluff rules are generally designed to standardize and unify SQL formatting/style conventions among valid syntactic alternatives, rather than acting as a deep semantic or syntax validator.
